core/kotlinx-coroutines-core/src/internal/CustomServiceLoader.kt

```diff
@@ -0,0 +1,116 @@
+package kotlinx.coroutines.internal
+
+import java.io.*
+import java.lang.ClassCastException
+import java.net.URL
+import java.util.*
+import java.util.jar.JarFile
+import java.util.zip.ZipEntry
+import kotlin.NoSuchElementException
+import java.io.IOException
+import java.io.BufferedReader
+
+private const val PREFIX: String = "META-INF/services/"
+
+public fun <S> loadProviders(service: Class<S>, loader: ClassLoader) = CustomServiceLoader(service, loader)
+
+public class CustomServiceLoader<S> internal constructor (private val service: Class<S>, private val loader: ClassLoader?) : Iterable<S> {
+
+    private fun parse(service: Class<S>, url: URL): Iterator<String> {
+        val names = mutableListOf<String>()
+        try {
+            val string = url.toString()
+            val separatorIndex = string.indexOf('!')
+            val pathToJar = string.substring(0, separatorIndex)
+            val entry = string.substring(separatorIndex + 2, string.length)
+            (JarFile(pathToJar.substring("jar:file:/".length), false) as Closeable).use { file ->
+                BufferedReader(InputStreamReader((file as JarFile).getInputStream(ZipEntry(entry)), "UTF-8")).use { r ->
+                    var lineCounter = 1
+                    while (lineCounter >= 0) {
+                        lineCounter = parseLine(service, url, r, lineCounter, names)
+                    }
+                }
+            }
+        } catch (e: Throwable) {
+            error(service, "Error reading configuration file", e)
+        }
+        return names.iterator()
+    }
+
+    private fun parseLine(service: Class<*>, u: URL, r: BufferedReader, lc: Int, names: MutableList<String>): Int {
+        var line = r.readLine() ?: return -1
+        val commentPos = line.indexOf('#')
+        if (commentPos >= 0) line = line.substring(0, commentPos)
+        line = line.trim()
+        val n = line.length
+        if (n != 0) {
+            if (line.indexOf(' ') >= 0 || line.indexOf('\t') >= 0)
+                error(service, "$u : $lc : Illegal configuration-file syntax")
+            var cp = line.codePointAt(0)
+            if (!Character.isJavaIdentifierStart(cp))
```

Comment:
Are all these checks really required? Something like `require(line.all {Character. isJavaIdentifierPart(it) })` should be enough.
Also, we can assume we have a single service per file and simplify it even more